File: sources/academy/webscraping/scraping_basics_python/exercises/guardian_f1_links.py
import httpx
from bs4 import BeautifulSoup
from urllib.parse import urljoin

listing_url = "https://www.theguardian.com/sport/formulaone"
response = httpx.get(listing_url)
response.raise_for_status()

soup = BeautifulSoup(response.text, "html.parser")

for item in soup.select('#maincontent ul li'):
    link = item.select_one('a')
    if link and 'href' in link.attrs:
        url = urljoin(listing_url, link['href'])
        print(url)

File: sources/academy/webscraping/scraping_basics_javascript/exercises/guardian_f1_links.mjs
import * as cheerio from 'cheerio';

const listingUrl = 'https://www.theguardian.com/sport/formulaone';
const response = await fetch(listingUrl);

if (!response.ok) {
  throw new Error(`HTTP ${response.status}`);
}

const html = await response.text();
const $ = cheerio.load(html);

for (const element of $('#maincontent ul li').toArray()) {
  const $item = $(element);
  const $link = $item.find('a').first();
  if ($link.length) {
    const url = new URL($link.attr('href'), listingUrl).href;
    console.log(url);
  }
}
